### PR TITLE
Runtime: Fix problems with nil-to-nil casts

### PR DESCRIPTION
**Pull request by slavapestov**
This fixes a crash regression from this patch:

https://github.com/apple/swift/commit/35cb1afab8b5009604cd4405375a3931d9f53f01

#### Resolved bug number: ([SR-1056](https://bugs.swift.org/browse/SR-1056))

---

Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

We have a special rule that Optional<T>.none successfully dynamically casts
to Optional<U>.none for any T and U. However the implementation was incorrect
if the source and destination types had a different size. We would initialize
the source to nil, and then copy to the result.

The correct implementation is to initialize the result using the result
payload type directly, and not call _succeed() at all.

Fixes https://bugs.swift.org/browse/SR-1056.

### CODE REVIEW

**Comment by slavapestov** [comment]
@jckarter and @tkremenek Can you please take a look?

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by gribozavr** [comment]
> We have a special rule that Optional.none successfully dynamically casts to Optional.none for any T and U.

Not for 2.2, but this rule seems to be a special case and inconsistent with the static type system.  Usually, as far as I understand, we try to make dynamic casts resemble the static type system.  Do you know why we have this behavior?  Should we consider removing this behavior?

**Comment by slavapestov** [comment]
The build failed with this, anyone know why?

    ld: library not found for -lswiftDarwin for architecture x86_64

**Comment by gribozavr** [comment]
Looks like we need to cherry-pick a609277f55e259b5ceb3fe19530a821c65df5df0 into swift-2.2-branch.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

File: test/1_stdlib/Casts.swift [hunk 1]
```diff
@@ -28,30 +28,9 @@ import ObjectiveC
...
-}
-
-func testFailedTupleCast(onDeinit: () -> ()) {
-    // This function is to establish a scope for t to 
-    // be deallocated at the end of.
-    let t: Any = (1, DeinitTester(onDeinit: onDeinit))
-    _ = t is Any.Type
-}
-
 CastsTests.test("No leak for failed tuple casts") {
```
Comment (gparker42) [inline test/1_stdlib/Casts.swift]: Is this test change part of your fix, or is it an unrelated cleanup?

Comment (slavapestov) [inline test/1_stdlib/Casts.swift]: As part of the fix I made the existing cast tests more comprehensive. And as part of doing that, I decided to rip out the duplicated 'DeinitTester' and use the standard LifetimeTracker mechanism.

**Comment by jckarter** [comment]
LGTM.

**Comment by slavapestov** [comment]
Unfortunately it's not quite ready to go in. There's some other difference in behavior between 2.2 and 3.0 and I'm trying to figure out if its a regression or expected.

**Comment by slavapestov** [comment]
Looks like I need to cherry-pick https://github.com/apple/swift/pull/1799 too. I'll go ahead and do that unless there are objections.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
Hey @atrick and @tkremenek, can you sign off on the inclusion of the other patch in this PR?

**Comment by atrick** [comment]
@slavapestov Yes, please include the patch for SR-912, #1799. It's quite simple, I just didn't know if it was worth taking into 2.2.

**Comment by tkremenek** [comment]
This appears to have broken the iOS simulator tests:

```
******************** TEST 'Swift :: 1_stdlib/Optional.swift' FAILED ********************
Script:
--
rm -rf /Users/buildnode/jenkins/workspace/oss-swift-2.2_tools-RA_stdlib-RDA_test-simulator/Ninja-ReleaseAssert/swift-macosx-x86_64/test-macosx-x86_64/1_stdlib/Output/Optional.swift.tmp && mkdir -p /Users/buildnode/jenkins/workspace/oss-swift-2.2_tools-RA_stdlib-RDA_test-simulator/Ninja-ReleaseAssert/swift-macosx-x86_64/test-macosx-x86_64/1_stdlib/Output/Optional.swift.tmp && xcrun --toolchain default --sdk /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk /Users/buildnode/jenkins/workspace/oss-swift-2.2_tools-RA_stdlib-RDA_test-simulator/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/swiftc -target x86_64-apple-macosx10.9  -module-cache-path '/var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/swift-testsuite-clang-module-cacherHXap5' -F /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/../../../Developer/Library/Frameworks -Xlinker -rpath -Xlinker /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk/../../../Developer/Library/Frameworks   -O -module-cache-path '/var/folders/_8/79jmzf2142z2xydc_01btlx00000gn/T/swift-testsuite-clang-module-cacherHXap5' /Users/buildnode/jenkins/workspace/oss-swift-2.2_tools-RA_stdlib-RDA_test-simulator/swift/test/1_stdlib/Optional.swift -o /Users/buildnode/jenkins/workspace/oss-swift-2.2_tools-RA_stdlib-RDA_test-simulator/Ninja-ReleaseAssert/swift-macosx-x86_64/test-macosx-x86_64/1_stdlib/Output/Optional.swift.tmp/a.out -module-name main &&  /Users/buildnode/jenkins/workspace/oss-swift-2.2_tools-RA_stdlib-RDA_test-simulator/Ninja-ReleaseAssert/swift-macosx-x86_64/test-macosx-x86_64/1_stdlib/Output/Optional.swift.tmp/a.out
--
Exit Code: 134

Command Output (stdout):
--
[ RUN      ] Optional.nil comparison
[       OK ] Optional.nil comparison
[ RUN      ] Optional.Equatable
[       OK ] Optional.Equatable
[ RUN      ] Optional.initializers
[       OK ] Optional.initializers
[ RUN      ] Optional.nil comparison
[       OK ] Optional.nil comparison
[ RUN      ] Optional.??
[       OK ] Optional.??
[ RUN      ] Optional.flatMap
[       OK ] Optional.flatMap
[ RUN      ] Optional.Casting Optional
```

https://ci.swift.org/view/swift-2.2-branch/job/oss-swift-2.2_tools-RA_stdlib-RDA_test-simulator/5269/consoleFull#-2021626272fca400bf-2f4a-462e-b517-e058d770b2d7

**Comment by tkremenek** [comment]
Reverted in #1985

**Comment by slavapestov** [comment]
The behavior change was there all along between optimized and non-optimized builds. The problem is with -O, the optimizer specializes the cast functions (even though they’re @inline(never)) so the runtime does not actually get tested in that case.
